Non-null assertion doesn't work for mapped types

<!-- 🚨 STOP 🚨 𝗦𝗧𝗢𝗣 🚨 𝑺𝑻𝑶𝑷 🚨 -->
<!--
Half of all issues filed here are duplicates, answered in the FAQ, or not appropriate for the bug tracker.

Please help us by doing the following steps before logging an issue:
  * Search: https://github.com/Microsoft/TypeScript/search?type=Issues
  * Read the CONTRIBUTING guidelines: https://github.com/Microsoft/TypeScript/blob/master/CONTRIBUTING.md
  * Read the FAQ: https://github.com/Microsoft/TypeScript/wiki/FAQ
-->

<!-- If you have a QUESTION:
   THIS IS NOT A FORUM FOR QUESTIONS.
   Ask questions at http://stackoverflow.com/questions/tagged/typescript
    or https://gitter.im/Microsoft/TypeScript
-->

<!-- If you have a SUGGESTION:
  Most suggestion reports are duplicates, please search extra hard before logging a new suggestion.
  See https://github.com/Microsoft/TypeScript-wiki/blob/master/Writing-Good-Design-Proposals.md
-->

<!-- If you have a BUG:
  Please fill in the *entire* template below.
-->

<!-- Please try to reproduce the issue with `typescript@next`. It may have already been fixed. -->
**TypeScript Version:** 
Reproducible in 2.7.2 and 2.9.0-dev.20180502.
Irreproducible in 2.6.2

<!-- Search terms you tried before logging this (so others can find this issue more easily) -->
**Search Terms:** 
Mapped types, Non-null assertion

**Code**

```ts
function f<A extends string>(p0: {[key in A]: {}|undefined}, p1: A) {
  const v: {} = p0[p1]!;
}
```

**Expected behavior:**
No errors.

**Actual behavior:**
In 2.7.2:
error TS2322: Type '{ [key in A]: {} | undefined; }[A]' is not assignable to type '{}'.
Type '{} | undefined' is not assignable to type '{}'.
Type 'undefined' is not assignable to type '{}'.

In 2.9.0-dev.20180502:
error TS2322: Type 'NonNullable<{ [key in A]: {} | undefined; }[A]>' is not assignable to type '{}'.
Type '{} | undefined' is not assignable to type '{}'.
Type 'undefined' is not assignable to type '{}'.

**Playground Link:** <!-- A link to a TypeScript Playground "Share" link which demonstrates this behavior --> 
I cannot reproduce the error in the playground.
From command line, I'm running `node_modules/.bin/tsc -strict test.ts` to reproduce it.

**Related Issues:**
https://github.com/Microsoft/TypeScript/issues/19608

@rkirov

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -6501,9 +6501,9 @@ namespace ts {
             // over the conditional type and possibly reduced. For example, 'T extends undefined ? never : T'
             // removes 'undefined' from T.
             if (type.root.isDistributive) {
-                const constraint = getConstraintOfType(type.checkType);
+                const constraint = getConstraintOfType(getSimplifiedType(type.checkType));
                 if (constraint) {
-                    const mapper = createTypeMapper([<TypeParameter>type.root.checkType], [constraint]);
+                    const mapper = makeUnaryTypeMapper(type.root.checkType, constraint);
                     const instantiated = getConditionalTypeInstantiation(type, combineTypeMappers(mapper, type.mapper));
                     if (!(instantiated.flags & TypeFlags.Never)) {
                         return instantiated;

--- a/tests/baselines/reference/nonNullMappedType.js
+++ b/tests/baselines/reference/nonNullMappedType.js
@@ -1,0 +1,10 @@
+//// [nonNullMappedType.ts]
+function f<A extends string>(p0: { [key in A]: {} | undefined }, p1: A) {
+    const v: {} = p0[p1]!;
+}
+
+//// [nonNullMappedType.js]
+"use strict";
+function f(p0, p1) {
+    var v = p0[p1];
+}

--- a/tests/baselines/reference/nonNullMappedType.symbols
+++ b/tests/baselines/reference/nonNullMappedType.symbols
@@ -1,0 +1,15 @@
+=== tests/cases/compiler/nonNullMappedType.ts ===
+function f<A extends string>(p0: { [key in A]: {} | undefined }, p1: A) {
+>f : Symbol(f, Decl(nonNullMappedType.ts, 0, 0))
+>A : Symbol(A, Decl(nonNullMappedType.ts, 0, 11))
+>p0 : Symbol(p0, Decl(nonNullMappedType.ts, 0, 29))
+>key : Symbol(key, Decl(nonNullMappedType.ts, 0, 36))
+>A : Symbol(A, Decl(nonNullMappedType.ts, 0, 11))
+>p1 : Symbol(p1, Decl(nonNullMappedType.ts, 0, 64))
+>A : Symbol(A, Decl(nonNullMappedType.ts, 0, 11))
+
+    const v: {} = p0[p1]!;
+>v : Symbol(v, Decl(nonNullMappedType.ts, 1, 9))
+>p0 : Symbol(p0, Decl(nonNullMappedType.ts, 0, 29))
+>p1 : Symbol(p1, Decl(nonNullMappedType.ts, 0, 64))
+}

--- a/tests/baselines/reference/nonNullMappedType.types
+++ b/tests/baselines/reference/nonNullMappedType.types
@@ -1,0 +1,17 @@
+=== tests/cases/compiler/nonNullMappedType.ts ===
+function f<A extends string>(p0: { [key in A]: {} | undefined }, p1: A) {
+>f : <A extends string>(p0: { [key in A]: {} | undefined; }, p1: A) => void
+>A : A
+>p0 : { [key in A]: {} | undefined; }
+>key : key
+>A : A
+>p1 : A
+>A : A
+
+    const v: {} = p0[p1]!;
+>v : {}
+>p0[p1]! : NonNullable<{ [key in A]: {} | undefined; }[A]>
+>p0[p1] : { [key in A]: {} | undefined; }[A]
+>p0 : { [key in A]: {} | undefined; }
+>p1 : A
+}

--- a/tests/cases/compiler/nonNullMappedType.ts
+++ b/tests/cases/compiler/nonNullMappedType.ts
@@ -1,0 +1,4 @@
+// @strict: true
+function f<A extends string>(p0: { [key in A]: {} | undefined }, p1: A) {
+    const v: {} = p0[p1]!;
+}

